src/bitset.ml

```diff
@@ -0,0 +1,224 @@
+module Make (B : Context.A_DISK) = struct
+  module Sector = Sector.Make (B)
+  module Schema = Schema.Make (B)
+
+  type t = Sector.t
+
+  open Lwt_result.Syntax
+
+  let get_page_size () = B.page_size
+
+  let get_nb_leaves () =
+    let nb_sectors = Int64.to_int B.nb_sectors in
+    let page_size = get_page_size () in
+    let bit_size = page_size * 8 in
+    (nb_sectors + bit_size - 1) / bit_size
+
+  let get_group_size nb_children nb_leaves =
+    let rec get_group_size group_size =
+      if group_size * nb_children >= nb_leaves
+      then group_size
+      else get_group_size (group_size * nb_children)
+    in
+    get_group_size 1
+
+  let get_ptr_size () =
+    let pointer_size = Sector.ptr_size in
+    let id_size = Sector.id_size in
+    ((pointer_size + id_size) / 8) + 8
+
+  let get_nb_children page_size =
+    let incr = get_ptr_size () in
+    (page_size - 4) / incr (* 4 bytes for the sector id*)
+
+  let get value offset = value land (1 lsl offset)
+
+  let get_flag t i =
+    let pos = i / 8 in
+    let* value = Sector.get_uint8 t pos in
+    let offset = i mod 8 in
+    let flag = get value offset in
+    Lwt_result.return flag
+
+  let get_leaf t i =
+    let page_size = get_page_size () in
+    let bit_size = page_size * 8 in
+    let leaf_ind = i / bit_size in
+    let nb_leaves = get_nb_leaves () in
+    let nb_children = get_nb_children page_size in
+    let incr = get_ptr_size () in
+    let rec reach_leaf t leaf_ind nb_leaves =
+      let group_size = get_group_size nb_children nb_leaves in
+      let child_ind = leaf_ind / group_size in
+      let new_leaves =
+        if child_ind = (nb_leaves - 1) / group_size
+        then ((nb_leaves - 1) mod group_size) + 1
```

Comment:
I'm not sure I understand why it's necessary to subtract 1 from `nb_leaves` on those two lines hmhmm (but I may be wrong, it's super hard to reason about offsets by 1 :/ )

Reply:
Thanks for the explanation, it actually looks good!